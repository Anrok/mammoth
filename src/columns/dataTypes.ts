import {
  CitextColumn,
  Column,
  DateColumn,
  DecimalColumn,
  IntegerColumn,
  JSONBColumn,
  JSONColumn,
  SerialColumn,
  TextColumn,
  TimeColumn,
  TimestampColumn,
  TimestampWithTimeZoneColumn,
  TimestampWithoutTimeZoneColumn,
  UuidColumn,
} from '.';

export const dataType = <T>(dataType: string) => new Column<T>(dataType);
export const uuid = () => new UuidColumn();
export const text = <T = string>() => new TextColumn<T>();
export const citext = () => new CitextColumn();
export const caseInsensitiveText = () => new CitextColumn();
export const integer = () => new IntegerColumn();
export const decimal = () => new DecimalColumn();
export const serial = () => new SerialColumn();
export const json = () => new JSONColumn();
export const jsonb = () => new JSONBColumn();
export const timestampWithTimeZone = () => new TimestampWithTimeZoneColumn();
export const timestamptz = () => new TimestampWithTimeZoneColumn();
export const timestampWithoutTimeZone = () => new TimestampWithoutTimeZoneColumn();
export const timestamp = () => new TimestampColumn();
export const date = () => new DateColumn();
export const time = () => new TimeColumn();
