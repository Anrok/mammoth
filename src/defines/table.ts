// TODO: currently this is redudant but we want to add options here at some point. We also want to
// introduce a single set of functions to define tables and indices, enums, functions, etc.
export const defineTable = <T>(tableSchema: T): T => {
  return tableSchema;
};
