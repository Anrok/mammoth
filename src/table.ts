import { Column, toSnakeCase, ColumnWrapper } from './columns';

export class TableType {
  [columnName: string]: Column<any>;
}

export class Table<Row, InsertRow = Row, UpdateRow = Row> {
  private readonly $name: string;
  private readonly $columnNames: ReadonlyArray<keyof Row>;
  private readonly $userDefinedTable: TableType;

  $row: Row;
  $insertRow: InsertRow;
  $updateRow: UpdateRow;

  constructor(userDefinedTable: TableType, name: string) {
    this.$userDefinedTable = userDefinedTable;
    this.$name = toSnakeCase(name);
    this.$columnNames = Object.keys(userDefinedTable) as (keyof Row)[];

    const self = this as any;
    this.$columnNames.forEach(camelCaseName => {
      const column = userDefinedTable[camelCaseName as string];
      if (!(column instanceof Column)) {
        throw new Error(`Invalid column at ${name}#${camelCaseName}`);
      }

      const snakeCaseName = column.getSnakeCaseName(camelCaseName as string);

      if (self[camelCaseName]) {
        throw new Error(
          `Column \`${camelCaseName}\` in table \`${name}\` collides with property of the same name in TableWrapper class.`,
        );
      }

      // TODO: change ColumnWrapper to Column instead ?!
      self[camelCaseName] = new ColumnWrapper(this, column, camelCaseName as string, snakeCaseName);
    });

    this.$row = undefined as any;
    this.$insertRow = undefined as any;
    this.$updateRow = undefined as any;
  }

  init() {
    this.$columnNames.forEach(columnName => {
      const column = this.$userDefinedTable[columnName as string];

      column.createReference();
    });
  }

  /** @internal */
  getColumns() {
    return this.$columnNames.map(columnName => this.getColumn(columnName)!);
  }

  /** @internal */
  getColumn(columnName: string | symbol | number): ColumnWrapper<any, any, any, any, any> {
    return (this as any)[columnName];
  }

  /** @internal */
  getName() {
    return this.$name;
  }
}
