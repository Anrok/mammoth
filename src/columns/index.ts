import { Default, Keyword } from '../keywords';
import { PartialQuery } from '../query';
import { TableWrapper } from '../table';
import { CollectionToken, GroupToken, ParameterToken, StringToken, Token } from '../tokens';
import { Unsafe } from '../unsafe';

export interface ColumnConfig<T> {
	primary?: boolean;
	default?: T | string; // TODO: Instead of string this should be new Raw() or something.
	check?: string;
	notNull?: boolean;
	unique?: boolean;
	references?: {
		tableName: string;
		columnName: string;
	};
	columnFunction?: ColumnFunction<T, any>;
}
export type ColumnFunction<T, Db> = (db: Db) => Column<T, any, any, any>;

export type AggregateType = 'COUNT' | 'MIN' | 'MAX' | 'AVG' | 'SUM';

export class Column<T, IT = T | null, ST = T | null, UT = T> {
	dataType: string;

	/** @internal */
	type: T;
	/** @internal */
	insertType: IT;
	/** @internal */
	selectType: ST;
	/** @internal */
	updateType: UT;
	/** @internal */
	name?: string;
	/** @internal */
	key?: string;
	/** @internal */
	table?: TableWrapper<any>;

	config: ColumnConfig<T>;

  constructor(name?: string, config = {}) {
		this.name = name;

		// tslint:disable-next-line:no-object-literal-type-assertion
		this.type = {} as T;
		// tslint:disable-next-line:no-object-literal-type-assertion
		this.insertType = {} as IT;
		// tslint:disable-next-line:no-object-literal-type-assertion
		this.selectType = {} as ST;
		// tslint:disable-next-line:no-object-literal-type-assertion
		this.updateType = {} as UT;

		this.config = config;
	}

	/** @internal */
	getName() {
		// TODO: this should be a string, not string | undefined.
		return this.name;
	}

	/** @internal */
	setName(name: string) {
		this.name = name;
	}

	/** @internal */
	setKey(key: string) {
		this.key = key;
	}

	/** @internal */
  setTable<TW extends TableWrapper<any>>(table: TW) {
		// TODO: why do we need the complete table here? If it's just for the table's name, can we just set the name instead?
    this.table = table;
	}

	primary(): Column<T, T | null, T> {
		this.config.primary = true;

		return this as any as Column<T, T | null, T>;
	}

	primaryKey() {
		return this.primary();
	}

	unique() {
		this.config.unique = true;
		return this;
	}

	notNull(): Column<T, T, T> {
		this.config.notNull = true;

		return this as any as Column<T, T, T>;
	}

	check(sql: string) {
		this.config.check = sql;
		return this;
	}

	default(sql: T | Unsafe | Keyword): Column<T, T | null, ST, UT | Default> {
		this.config.default = (sql && (sql instanceof Unsafe || sql instanceof Keyword))
			? sql.toSql()
			: sql;
		return this as any as Column<T, T | null, ST, UT | Default>;
	}

	/** @internal */
	createReference<Db>(db: Db) {
		if (this.config.columnFunction) {
			const column = this.config.columnFunction(db);

			this.config.references = {
				tableName: column.table!.getName(),
				columnName: column.name!,
			};
			this.config.columnFunction = undefined;
		}
	}

	references<Db>(columnFunction: ColumnFunction<T, Db>) {
		this.config.columnFunction = columnFunction;
		return this;
	}

	// TODO: Move everything below this comment to a separate class e.g. QueryableColumn.
	// That way we can separate the schema-related methods and the query-related methods.

	private compare(value: T | Column<any> | PartialQuery, comparator: string) {
		const query = new PartialQuery();
		query.tokens.push(
			new StringToken(this.toSql()),
			new StringToken(comparator),
		);

		if (value instanceof Column) {
			query.tokens.push(new StringToken(`${value.table!.getName()}.${value.name}`));
		}
		else if (value instanceof PartialQuery) {
			query.tokens.push(...value.tokens);
		}
		else {
			query.tokens.push(new ParameterToken(value));
		}
		return query;
	}

	isNull() {
		const query = new PartialQuery();
		query.tokens.push(
			new StringToken(this.toSql()),
			new StringToken(`IS NULL`),
		);
		return query;
	}

	isNotNull() {
		const query = new PartialQuery();
		query.tokens.push(
			new StringToken(this.toSql()),
			new StringToken(`IS NOT NULL`),
		);
		return query;
	}

	asc() {
		return new PartialQuery(new StringToken(this.toSql()), new StringToken(`ASC`));
	}

	desc() {
		return new PartialQuery(new StringToken(this.toSql()), new StringToken(`DESC`));
	}

	in(array: T[]) {
		return new PartialQuery(
			new StringToken(`${this.table!.getName()}.${this.name}`),
			new StringToken(`IN`),
			new GroupToken([new ParameterToken(array)]),
		)
	}

	private aggregate(aggregateType: AggregateType) {
		// TODO: the selectType probably needs an update depending on the aggregate type?

		const aggregateColumn = new AggregateColumn<T, IT, ST, UT>(aggregateType, this.name!, this.key!);

		// TODO: this is a bit ugly. Is there anything else we need to set?
		aggregateColumn.table = this.table;
		return aggregateColumn;
	}

	toSql() {
		return `${this.table!.getName()}.${this.name}`;
	}

	toTokens(): Token[] {
		return [
			new StringToken(this.toSql()),
		];
	}

	count() { return this.aggregate(`COUNT`); }
	sum() { return this.aggregate(`SUM`); }
	min() { return this.aggregate(`MIN`); }
	max() { return this.aggregate(`MAX`); }
	avg() { return this.aggregate(`AVG`); }

	private operate(operator: '+' | '-' | '*' | '/' | '%', value: T) {
		return new PartialQuery(
			new StringToken(`${this.table!.getName()}.${this.name}`),
			new StringToken(operator),
			new ParameterToken(value),
		);
	}

	// TODO: should we add these on specific column types only? E.g. NumberColumn.
	plus(value: T) { return this.operate(`+`, value); }
	minus(value: T) { return this.operate(`+`, value); }
	multiply(value: T) { return this.operate(`+`, value); }
	divide(value: T) { return this.operate(`/`, value); }
	modulo(value: T) { return this.operate(`%`, value); }

	// TODO: should this only be on TextColumn exclusively?
	concat(value: T | Column<any>) {
		return new PartialQuery(
			new StringToken(this.toSql()),
			new StringToken(`||`),
			value instanceof Column
				? new CollectionToken(value.toTokens())
				: new ParameterToken(value),
		);
	}

	between(a: T, b: T) {
		return new PartialQuery(
			new StringToken(this.toSql()),
			new StringToken(`BETWEEN`),
			new ParameterToken(a),
			new StringToken(`AND`),
			new ParameterToken(b),
		);
	}

	eq(value: T | Column<any> | PartialQuery) { return this.compare(value, `=`); }
	ne(value: T | Column<any> | PartialQuery) { return this.compare(value, `!=`); }
	gt(value: T | Column<any> | PartialQuery) { return this.compare(value, `>`); }
	gte(value: T | Column<any> | PartialQuery) { return this.compare(value, `>=`); }
	lt(value: T | Column<any> | PartialQuery) { return this.compare(value, `>=`); }
	lte(value: T | Column<any> | PartialQuery) { return this.compare(value, `<=`); }
}

export class AggregateColumn<T, IT, ST, UT> extends Column<T, IT, ST, UT> {
	aggregateType: string;

	constructor(aggregateType: string, name: string, key: string) {
		super();

		this.aggregateType = aggregateType;
		this.name = name;
		this.key = key;
	}

	toTokens() {
		return [
			new GroupToken(
				[new StringToken(this.toSql())],
				`${this.aggregateType}(`,
				`)`,
			),
		];
	}
}

export class TextColumn extends Column<string> {
	dataType = 'TEXT';
}
export class CitextColumn extends Column<string> {
	dataType = 'CITEXT';
}
export class CaseInsensitiveTextColumn extends CitextColumn {}
export class IntegerColumn extends Column<number> {
	dataType = 'INTEGER';
}
export class DecimalColumn extends Column<number> {
	dataType = 'DECIMAL';
}
export class SerialColumn extends Column<number, number | null> {
	dataType = 'SERIAL';
}
export class JSONColumn<T> extends Column<T> {
	dataType = 'JSON';
}
export class JSONBColumn<T> extends Column<T> {
	dataType = 'JSONB';
}
export class TimestampWithTimeZoneColumn extends Column<Date> {
	dataType = 'TIMESTAMP WITH TIME ZONE';
}
export class TimestampWithoutTimeZoneColumn extends Column<Date> {
	dataType = 'TIMESTAMP WITHOUT TIME ZONE';
}

export type Uuid = string;

export class UuidColumn extends Column<Uuid> {
	dataType = 'UUID';
}

export class StringColumn extends TextColumn {}
export class NumberColumn extends IntegerColumn {}
export class DateColumn extends TimestampWithTimeZoneColumn {}