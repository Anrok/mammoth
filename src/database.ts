import * as pg from 'pg';
import * as url from 'url';
import { Column } from './columns';
import { PartialQuery, Query, Tokenable } from './query';
import { Table, TableWrapper } from './table';
import { GroupToken, StringToken } from './tokens';
import { Transaction } from './transaction';
import { Unsafe } from './unsafe';

// FIXME: any should be replaced by something specific. But specifying Table, which should be the
// right type, breaks functionality. Table adds an index signature which means doing keyof MyTable
// (which extends from Table) returns string instead a type with all keys.
export interface TableMap { [tableName: string]: any }

const createTable = <T extends Table>(table: T, tableName: string) => {
  type Row = { [P in keyof T]: T[P]['selectType'] };
  type InsertRow = { [P in keyof T]: T[P]['insertType'] };
  type UpdateRow = { [P in keyof T]: T[P]['updateType'] };

  const newTable = new TableWrapper<Row, InsertRow, UpdateRow>(table, tableName);

  return newTable as T & TableWrapper<Row, InsertRow, UpdateRow>;
}

export interface DatabaseConfig {
  min?: number;
  max?: number;
}

export class Database<Tables> {
  pool: pg.Pool;
  tableNames: string[];

  private createPool(databaseUrl: string, options: DatabaseConfig): pg.Pool {
		const { auth, hostname, port, pathname } = url.parse(databaseUrl);
		const [ user, password ] = (auth || '').split(':');

		const config = {
			...options,
			user,
			password,
			host: hostname,
			port: parseInt(port || '5432', 10),
			database: (pathname || '').slice(1),
			ssl: process.env.DB_DISABLE_SSL !== 'true',
    };

		return new pg.Pool(config);
  }

  constructor(tableNames: string[]) {
    if (!process.env.DATABASE_URL) {
      throw new Error(`DATABASE_URL is not set.`);
    }

    this.pool = this.createPool(String(process.env.DATABASE_URL), {
      min: process.env.DB_MIN_POOL_SIZE
        ? parseInt(process.env.DB_MIN_POOL_SIZE!, 10)
        : undefined,
      max: process.env.DB_MAX_POOL_SIZE
        ? parseInt(process.env.DB_MAX_POOL_SIZE!, 10)
        : undefined,
    });
    this.tableNames = tableNames;
  }

  sql(strings: TemplateStringsArray, ...parameters: any[]) {
    const text = strings.reduce((query, string, index) => query + string + (index < parameters.length
      ? `$${String(index + 1)}`
      : ``), ``);

    return this.exec(text, parameters);
  }

  async transaction(callback: (transaction: Transaction & Tables) => Promise<any> | void) {
    const client = await this.pool.connect();
    const transaction: Transaction & Tables = new Transaction(client) as any;

    try {
      await transaction.begin();

      await Promise.resolve(callback(transaction));

      await transaction.commit();

      client.release();
    }
    catch (e) {
      try {
        await transaction.rollback();
      }
      catch (e2) {
        // If rollback fails we assume this client is bugged and we don't want to give it back
        // to the pool. By passing in the error we effectively tell the pool to disconnect and
        // forget about the client.
        client.release(e2);
      }

      client.release();
    }
  }

  async exec(text: string, parameters: any[] = []) {
    let client: pg.Client | undefined;

		try {
			client = await this.pool.connect();

      const result = await client.query(text, parameters);

			return result;
		}
		catch (e) {
      e.query = text;

			throw e;
		}
		finally {
			if (client) {
				client.release();
			}
		}
  }

  now() {
    const query = new PartialQuery();
    query.tokens.push(new StringToken(`NOW()`));
    return query;
  }

  default() {
    return new Default();
  }

  uuidGenerateV4() {
    return new Unsafe(`uuid_generate_v4()`);
  }

  excluded(_column: Column<any>) {
    //
  }

  exists(_query: Query<any, any, any, any>) {
    //
  }

  not(tokenable: Tokenable) {
    return new PartialQuery(
      new StringToken(`NOT`),
      new GroupToken(tokenable.tokens),
    );
  }
}

export abstract class Keyword {
  abstract toString(): string;

  toTokens() {
    return [
      new StringToken(this.toString()),
    ];
  }
}
export class Default extends Keyword {
  toString() { return `DEFAULT`; }
}

export class UuidGenerateV4 extends Keyword {
  toString() { return `uuid_generate_v4()`; }
}

export class Now extends Keyword {
  toString() { return `NOW()`; }
}

export class CurrentTimestamp extends Keyword {
  toString() { return `CURRENT_TIMESTAMP`; }
}

export const createDatabase = <T extends TableMap>(tables: T) => {
  const tableNames = Object.keys(tables) as (keyof T)[];

  type State = {
    [P in keyof T]: T[P] & TableWrapper<{
      [C in keyof T[P]]: T[P][C]['selectType'];
    }, {
      [C in keyof T[P]]: T[P][C]['insertType'];
    }, {
      [C in keyof T[P]]: T[P][C]['updateType'];
    }>
  };

  const database: Database<State> & State = new Database<State>(tableNames) as any;

  tableNames.forEach(tableName => {
    const table = tables[tableName];

    database[tableName] = createTable(table, tableName);
  });

  tableNames.forEach(tableName => {
    const table: TableWrapper<any> = database[tableName];
    table.init(database);
  });

  return database as Database<State> & State;
}