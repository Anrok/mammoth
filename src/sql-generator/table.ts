import { TableWrapper } from '../table';
import { EnumColumn } from '..';

// TODO: We could also immediately generate the AST from here?

export const generateSql = <T extends TableWrapper<any>>(table: T) => {
  return [...generateTypeSql(table), generateCreateTableSql(table)];
};

export const generateTypeSql = <T extends TableWrapper<any>>(table: T) => {
  return table
    .getColumns()
    .filter(column => column.column instanceof EnumColumn)
    .map(column => {
      const enumColumn = column.column as EnumColumn<any>;

      if (!enumColumn.dataType) {
        enumColumn.dataType = `${table.getName()}_${column.snakeCaseName}_ENUM`.toUpperCase();
      }

      return `CREATE TYPE ${enumColumn.dataType} AS ENUM (${enumColumn.values
        .map((value: string) => `'${value}'`)
        .join(`, `)})`;
    });
};

export const generateCreateTableSql = <T extends TableWrapper<any>>(table: T) => {
  const columns = table.getColumns().map(column => {
    const config = column.getConfig();

    const parts = [
      `  ${column.snakeCaseName}`,

      // TODO: it's probably nicer to have a function here so once we refactor the original column
      // out of the column wrapper this still works.
      column.column.dataType,
    ];

    if (config.primary) {
      parts.push(`PRIMARY KEY`);
    }

    if (config.notNull) {
      parts.push(`NOT NULL`);
    }

    if (config.default !== undefined) {
      // TODO: should we escape this?
      parts.push(`DEFAULT ${config.default}`);
    }

    if (config.check) {
      // TODO: should we escape this?
      parts.push(`CHECK (${config.check})`);
    }

    if (config.unique) {
      parts.push(`UNIQUE`);
    }

    if (config.references) {
      parts.push(`REFERENCES ${config.references.tableName} (${config.references.columnName})`);

      if (config.onUpdate) {
        if (config.onUpdate === 'cascade') {
          parts.push(`ON UPDATE CASCADE`);
        } else if (config.onUpdate === 'restrict') {
          parts.push(`ON UPDATE RESTRICT`);
        } else {
          //
        }
      }

      if (config.onDelete) {
        if (config.onDelete === 'cascade') {
          parts.push(`ON DELETE CASCADE`);
        } else if (config.onDelete === 'restrict') {
          parts.push(`ON DELETE RESTRICT`);
        } else {
          //
        }
      }
    }

    return parts.join(' ');
  });

  return `CREATE TABLE ${table.getName()} (\n${columns.join(`,\n`)}\n)`;
};
