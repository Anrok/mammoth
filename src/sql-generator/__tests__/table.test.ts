import {
  CitextColumn,
  IntegerColumn,
  DecimalColumn,
  SerialColumn,
  JSONColumn,
  JSONBColumn,
  TimestampWithTimeZoneColumn,
  TimestampWithoutTimeZoneColumn,
  TimestampColumn,
  DateColumn,
  TimeColumn,
  TimeWithoutTimeZoneColumn,
  TimeWithTimeZoneColumn,
  IntervalColumn,
  MoneyColumn,
  BooleanColumn,
  ByteaColumn,
} from './../../columns';
import { Unsafe, UuidGenerateV4 } from '../..';
import { TextColumn, UuidColumn, EnumColumn } from '../../columns';
import { createDatabase } from '../../database/pool';
import { Table } from '../../table';
import { generateSql } from '../table';

const test = (clazz: any) => {
  const db = createDatabase({
    test: clazz,
  });

  return generateSql(db.test);
};

describe('generate sql', () => {
  it('should generate a simple table', () => {
    class Organization extends Table {
      id = new UuidColumn()
        .primary()
        .notNull()
        .default(new Unsafe(`uuid_generate_v4()`));
      apiKey = new TextColumn().notNull().unique();
    }

    const sql = test(new Organization());

    expect(sql).toMatchInlineSnapshot(`
                                    Array [
                                      "CREATE TABLE test (
                                      id UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
                                      api_key TEXT NOT NULL UNIQUE
                                    )",
                                    ]
                        `);
  });

  it('should generate a table with enum without name', () => {
    class EnumTest {
      id = new UuidColumn()
        .primaryKey()
        .notNull()
        .default(new UuidGenerateV4());
      value = new EnumColumn([`A`, `B`]).notNull();
    }

    const sql = test(new EnumTest());

    expect(sql).toMatchInlineSnapshot(`
                                    Array [
                                      "CREATE TYPE TEST_VALUE_ENUM AS ENUM ('A', 'B')",
                                      "CREATE TABLE test (
                                      id UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
                                      value TEST_VALUE_ENUM NOT NULL
                                    )",
                                    ]
                        `);
  });

  it('should generate a table with enum with name', () => {
    class EnumTest {
      id = new UuidColumn()
        .primaryKey()
        .notNull()
        .default(new UuidGenerateV4());
      value = new EnumColumn([`A`, `B`], `MY_ENUM`).notNull();
    }
    const sql = test(new EnumTest());

    expect(sql).toMatchInlineSnapshot(`
                                    Array [
                                      "CREATE TYPE MY_ENUM AS ENUM ('A', 'B')",
                                      "CREATE TABLE test (
                                      id UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
                                      value MY_ENUM NOT NULL
                                    )",
                                    ]
                        `);
  });

  it(`should create table with check`, () => {
    class Foo {
      id = new UuidColumn().primaryKey().notNull();
      bar = new TextColumn().check(`bar <> 'test'`);
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
                              Array [
                                "CREATE TABLE test (
                                id UUID PRIMARY KEY NOT NULL,
                                bar TEXT CHECK (bar <> 'test')
                              )",
                              ]
                    `);
  });

  it(`should create table with references`, () => {
    class Foo {
      id = new UuidColumn().primaryKey().notNull();
      bar = new UuidColumn().references((db: Db) => db.bar.id);
    }
    class Bar {
      id = new UuidColumn().primaryKey().notNull();
    }

    const db = createDatabase({
      foo: new Foo(),
      bar: new Bar(),
    });
    type Db = typeof db;

    const sql = generateSql(db.foo);
    expect(sql).toMatchInlineSnapshot(`
                              Array [
                                "CREATE TABLE foo (
                                id UUID PRIMARY KEY NOT NULL,
                                bar UUID REFERENCES bar (id)
                              )",
                              ]
                    `);
  });

  it(`should create table with references on update cascade`, () => {
    class Foo {
      id = new UuidColumn().primaryKey().notNull();
      bar = new UuidColumn()
        .references((db: Db) => db.bar.id)
        .onUpdate()
        .cascade();
    }
    class Bar {
      id = new UuidColumn().primaryKey().notNull();
    }

    const db = createDatabase({
      foo: new Foo(),
      bar: new Bar(),
    });
    type Db = typeof db;

    const sql = generateSql(db.foo);
    expect(sql).toMatchInlineSnapshot(`
                        Array [
                          "CREATE TABLE foo (
                          id UUID PRIMARY KEY NOT NULL,
                          bar UUID REFERENCES bar (id) ON UPDATE CASCADE
                        )",
                        ]
                `);
  });

  it(`should create table with references on update restrict`, () => {
    class Foo {
      id = new UuidColumn().primaryKey().notNull();
      bar = new UuidColumn()
        .references((db: Db) => db.bar.id)
        .onUpdate()
        .restrict();
    }
    class Bar {
      id = new UuidColumn().primaryKey().notNull();
    }

    const db = createDatabase({
      foo: new Foo(),
      bar: new Bar(),
    });
    type Db = typeof db;

    const sql = generateSql(db.foo);
    expect(sql).toMatchInlineSnapshot(`
                        Array [
                          "CREATE TABLE foo (
                          id UUID PRIMARY KEY NOT NULL,
                          bar UUID REFERENCES bar (id) ON UPDATE RESTRICT
                        )",
                        ]
                `);
  });

  it(`should create table with references on delete cascade`, () => {
    class Foo {
      id = new UuidColumn().primaryKey().notNull();
      bar = new UuidColumn()
        .references((db: Db) => db.bar.id)
        .onDelete()
        .cascade();
    }
    class Bar {
      id = new UuidColumn().primaryKey().notNull();
    }

    const db = createDatabase({
      foo: new Foo(),
      bar: new Bar(),
    });
    type Db = typeof db;

    const sql = generateSql(db.foo);
    expect(sql).toMatchInlineSnapshot(`
                        Array [
                          "CREATE TABLE foo (
                          id UUID PRIMARY KEY NOT NULL,
                          bar UUID REFERENCES bar (id) ON DELETE CASCADE
                        )",
                        ]
                `);
  });

  it(`should create table with references on delete restrict`, () => {
    class Foo {
      id = new UuidColumn().primaryKey().notNull();
      bar = new UuidColumn()
        .references((db: Db) => db.bar.id)
        .onDelete()
        .restrict();
    }
    class Bar {
      id = new UuidColumn().primaryKey().notNull();
    }

    const db = createDatabase({
      foo: new Foo(),
      bar: new Bar(),
    });
    type Db = typeof db;

    const sql = generateSql(db.foo);
    expect(sql).toMatchInlineSnapshot(`
                        Array [
                          "CREATE TABLE foo (
                          id UUID PRIMARY KEY NOT NULL,
                          bar UUID REFERENCES bar (id) ON DELETE RESTRICT
                        )",
                        ]
                `);
  });

  it(`should create table with citext`, () => {
    class Foo {
      val = new CitextColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
                  Array [
                    "CREATE TABLE test (
                    val CITEXT
                  )",
                  ]
            `);
  });

  it(`should create table with integer column`, () => {
    class Foo {
      val = new IntegerColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val INTEGER
            )",
            ]
        `);
  });

  it(`should create table with decimal column`, () => {
    class Foo {
      val = new DecimalColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val DECIMAL
            )",
            ]
        `);
  });

  it(`should create table with serial column`, () => {
    class Foo {
      val = new SerialColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val SERIAL
            )",
            ]
        `);
  });

  it(`should create table with JSON column`, () => {
    class Foo {
      val = new JSONColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val JSON
            )",
            ]
        `);
  });

  it(`should create table with JSONB column`, () => {
    class Foo {
      val = new JSONBColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val JSONB
            )",
            ]
        `);
  });

  it(`should create table with TimestampWithTimeZone column`, () => {
    class Foo {
      val = new TimestampWithTimeZoneColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val TIMESTAMP WITH TIME ZONE
            )",
            ]
        `);
  });

  it(`should create table with TimestampWithoutTimeZone column`, () => {
    class Foo {
      val = new TimestampWithoutTimeZoneColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val TIMESTAMP WITHOUT TIME ZONE
            )",
            ]
        `);
  });

  it(`should create table with timestamp column`, () => {
    class Foo {
      val = new TimestampColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val TIMESTAMP
            )",
            ]
        `);
  });

  it(`should create table with date column`, () => {
    class Foo {
      val = new DateColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val DATE
            )",
            ]
        `);
  });

  it(`should create table with time column`, () => {
    class Foo {
      val = new TimeColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val TIME
            )",
            ]
        `);
  });

  it(`should create table with time with time zone column`, () => {
    class Foo {
      val = new TimeWithTimeZoneColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val TIME WITH TIME ZONE
            )",
            ]
        `);
  });

  it(`should create table with time without time zone column`, () => {
    class Foo {
      val = new TimeWithoutTimeZoneColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val TIME WITHOUT TIME ZONE
            )",
            ]
        `);
  });

  it(`should create table with interval column`, () => {
    class Foo {
      val = new IntervalColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val INTERVAL
            )",
            ]
        `);
  });

  it(`should create table with money column`, () => {
    class Foo {
      val = new MoneyColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val MONEY
            )",
            ]
        `);
  });

  it(`should create table with boolean column`, () => {
    class Foo {
      val = new BooleanColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
            Array [
              "CREATE TABLE test (
              val BOOLEAN
            )",
            ]
        `);
  });

  it(`should create table with bytea column`, () => {
    class Foo {
      val = new ByteaColumn();
    }
    const sql = test(new Foo());
    expect(sql).toMatchInlineSnapshot(`
      Array [
        "CREATE TABLE test (
        val BYTEA
      )",
      ]
    `);
  });
});
