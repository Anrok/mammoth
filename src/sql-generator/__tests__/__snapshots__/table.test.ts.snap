// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate sql should generate a simple table 1`] = `
"CREATE TABLE organization (
  id UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
  api_key TEXT NOT NULL UNIQUE
)"
`;

exports[`generate sql should generate a table with enum with name 1`] = `
Array [
  "CREATE TYPE MY_ENUM AS ENUM ('A', 'B')",
  "CREATE TABLE enum_test (
  id UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
  value MY_ENUM NOT NULL
)",
]
`;

exports[`generate sql should generate a table with enum without name 1`] = `
Array [
  "CREATE TYPE ENUM_TEST_VALUE_ENUM AS ENUM ('A', 'B')",
  "CREATE TABLE enum_test (
  id UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
  value ENUM_TEST_VALUE_ENUM NOT NULL
)",
]
`;
